src/mainwindow.cpp

```diff
@@ -193,12 +198,35 @@ void MainWindow::openImageFile(QString fileName) {
     fm_path_unref(path);
     return;
   }
-  // load the image file asynchronously
-  loadImage(path);
-  loadFolder(fm_path_get_parent(path));
+  if (QFileInfo(fileName).isDir()) {
+    if(fm_path_equal(path, folderPath_) || !dirHasImage(fileName)) {
+      fm_path_unref(path);
+      return;
+    }
+    if(currentFile_)
+      fm_path_unref(currentFile_);
+    currentFile_ = nullptr;
+    loadFolder(path);
+  } else {
+    // load the image file asynchronously
+    loadImage(path);
+    loadFolder(fm_path_get_parent(path));
+  }
   fm_path_unref(path);
 }
 
+bool MainWindow::dirHasImage(QString dirname)
+{
+  QList<QByteArray> formats = QImageReader::supportedImageFormats();
+  QStringList formatsFilters;
+  for (const QByteArray& format: formats)
+    formatsFilters << QString("*.") + format;
+  QDir dir(dirname);
+  dir.setNameFilters(formatsFilters);
+  dir.setFilter(QDir::Files | QDir::NoDotAndDotDot);
+  return !dir.entryList().isEmpty();
+}
+
```

Comment:
I personnaly find it less readable to have all these operations packed into the same method.